plugins/auth-backend/src/lib/OAuthProvider.ts

```diff
@@ -39,14 +40,40 @@ export type Options = {
   tokenIssuer: TokenIssuer;
 };
 
+const readState = (stateString: string): OAuthState => {
+  try {
+    const state = Object.fromEntries(
+      new URLSearchParams(decodeURIComponent(stateString)),
+    );
+    return {
+      nonce: state.nonce ?? '',
+      env: state.env ?? '',
+    };
+  } catch (e) {
+    return {
+      nonce: '',
+      env: '',
+    };
+  }
+};
+
+export const encodeState = (state: OAuthState): string => {
+  const searchParams = new URLSearchParams();
+  searchParams.append('nonce', state.nonce ?? '');
+  searchParams.append('env', state.env ?? '');
+
+  return encodeURIComponent(searchParams.toString());
+};
+
 export const verifyNonce = (req: express.Request, providerId: string) => {
   const cookieNonce = req.cookies[`${providerId}-nonce`];
-  const stateNonce = req.query.state;
+  const state: OAuthState = readState(req.query.state?.toString() ?? '');
+  const stateNonce = state.nonce;
 
   if (!cookieNonce) {
     throw new Error('Auth response is missing cookie nonce');
   }
-  if (!stateNonce) {
+  if (stateNonce.length === 0) {
     throw new Error('Auth response is missing state nonce');
```

Comment:
This error could for example be moved in and thrown from the `readState` function instead.

Reply:
Done. I moved the empty string check into the `readState` function.